### PR TITLE
Continuous Gaussian Sample for MatQ

### PR DESCRIPTION
**Pull request by Marvin-Beckmann**
**Description**

This PR implements...
- [x] continuous Gaussian sampling for `MatQ`.
Here, each entry is sampled using `Q::sample_gauss`.

**Testing**

- [x] I added basic working examples (possibly in doc-comment)
- [x] I triggered all possible errors in my test in every possible way
- [x] I included tests for all reasonable edge cases

**Checklist:**

- [x] I have performed a self-review of my own code
  - [x] The code provides good readability and maintainability s.t. it fulfills best practices like talking code, modularity, ...
    - [x] The chosen implementation is not more complex than it has to be
  - [x] My code should work as intended and no side effects occur (e.g. memory leaks)
  - [x] The doc comments fit our style guide
  - [x] I have credited related sources if needed

### CODE REVIEW

File: src/rational/mat_q/sample/gauss.rs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    ///
+    /// Parameters:
+    /// - `num_rows`: specifies the number of rows the new matrix should have
+    /// - `num_cols`: specifies the number of columns the new matrix should have
+    /// - `center`: specifies the position of the center
+    /// - `sigma`: specifies the standard deviation
+    ///
+    /// Returns new [`MatQ`] sample chosen according to the specified continuous Gaussian
+    /// distribution or a [`MathError`] if the specified parameters were not chosen
+    /// appropriately, `sigma > 0`.
```
Comment (Marcel583) [inline src/rational/mat_q/sample/gauss.rs:30]:
```suggestion
    /// appropriately (`sigma > 0`).
```

File: src/rational/mat_q/sample/gauss.rs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    /// - Returns a [`MathError`] of type [`NonPositive`](MathError::NonPositive)
+    ///     if `sigma <= 0`.
+    ///
+    /// # Panics ...
+    /// - if the provided number of rows and columns are not suited to create a matrix.
+    ///     For further information see [`MatQ::new`].
+    pub fn sample_gauss(
+        num_rows: impl TryInto<i64> + Display,
+        num_cols: impl TryInto<i64> + Display,
+        center: impl Into<f64>,
```
Comment (Marcel583) [inline src/rational/mat_q/sample/gauss.rs:49]: In Q, you had 'center: impl Into\<Q\>', why is it here 'Into\<f64\>'?
Maybe change it to 'Into\<Q\>'.

**Review by Marcel583 — APPROVED** [review]

File: src/rational/mat_q/sample/gauss.rs (hunk 2 above)
Comment (Marvin-Beckmann) [reply src/rational/mat_q/sample/gauss.rs]: reworked this to take in a MatQ instead which indirectly provides the dimensions for the sample -> please check again

File: src/rational/mat_q/sample/gauss.rs [hunk 3]
```diff
@@ -6,55 +6,44 @@
...
     /// distribution or a [`MathError`] if the specified parameters were not chosen
-    /// appropriately, `sigma > 0`.
+    /// appropriately (`sigma > 0`).
     ///
     /// # Examples
     /// ```
     /// use qfall_math::rational::MatQ;
     ///
-    /// let sample = MatQ::sample_gauss(5, 5, 0, 1).unwrap();
+    /// let sample = MatQ::sample_gauss(&MatQ::new(5, 5), 1).unwrap();
```
[the remark this answers is not in the document]
Comment (LPorzenheim) [reply src/rational/mat_q/sample/gauss.rs:33]: I would assume that noone will use this function, as I have never seen the use case, where you sample a matrix with a different center for each entry. Furthermore, at first the function confused me, as I expected something vastly different from the name.

[the remark this answers is not in the document]
Comment (Marvin-Beckmann) [reply src/rational/mat_q/sample/gauss.rs:35]: added a separat function for the previous version

File: src/rational/mat_q/sample/gauss.rs [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    /// ```
+    /// use qfall_math::rational::MatQ;
+    ///
+    /// let sample = MatQ::sample_gauss(&MatQ::new(5, 5), 1).unwrap();
+    /// ```
+    ///
+    /// # Errors and Failures
+    /// - Returns a [`MathError`] of type [`NonPositive`](MathError::NonPositive)
+    ///     if `sigma <= 0`.
+    pub fn sample_gauss(center: &MatQ, sigma: impl Into<f64>) -> Result<MatQ, MathError> {
```
Comment (Marcel583) [inline src/rational/mat_q/sample/gauss.rs:41]:
```suggestion
    pub fn sample_gauss_different_centers(centers: &MatQ, sigma: impl Into<f64>) -> Result<MatQ, MathError> {
```

File: src/rational/mat_q/sample/gauss.rs [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    rational::{MatQ, Q},
+    traits::{GetEntry, GetNumColumns, GetNumRows, SetEntry},
+};
+
+impl MatQ {
+    /// Chooses a [`MatQ`] instance according to the continuous Gaussian distribution.
+    /// Here, each entry is chosen according to the provided distribution.
+    ///
+    /// Parameters:
+    /// - `center`: specifies the center for each entry of the matrix individually
```
Comment (Marcel583) [inline src/rational/mat_q/sample/gauss.rs:24]:
```suggestion
    /// - `centers`: specifies the center for each entry of the matrix individually
```

File: src/rational/mat_q/sample/gauss.rs [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    /// let sample = MatQ::sample_gauss_same_center(5, 5, &center, 1).unwrap();
+    /// ```
+    ///
+    /// # Errors and Failures
+    /// - Returns a [`MathError`] of type [`NonPositive`](MathError::NonPositive)
+    ///     if `sigma <= 0`.
+    ///
+    /// # Panics ...
+    /// - if the number of rows or columns is negative, `0`, or does not fit into an [`i64`].
+    pub fn sample_gauss_same_center(
```
Comment (Marcel583) [inline src/rational/mat_q/sample/gauss.rs:85]:
```suggestion
    pub fn sample_gauss(
```

**Review by Marcel583 — COMMENTED** [review]
Looks good, I would just rename the same center function to be the main function and the different center function to be the special one

File: src/rational/mat_q/sample/gauss.rs (hunk 6 above)
Comment (Marvin-Beckmann) [reply src/rational/mat_q/sample/gauss.rs:85]: talked about it with Laurens when I made the naming this way -> will leave it as it is

**Pull request merged**